src/macros-by-example.adoc

```diff
@@ -711,71 +711,62 @@ Thank you to my Patreon supporters:
    (currency *symbol*)]
   (*for* [(amt amount)]
     (price amt currency)))
 ----
 
-<1> Note the use of `[]` around `*number*`.
+Here, the parameter `amount` accepts any number of `*number*`s.
+It's easy to provide exactly one:
 
-This is referred to as a _grouped parameter_, and at invocation it requires a list delimiting its
-_argument group_:
+----
+(:prices 12.99 GBP) ⇒ {amount:12.99, currency:GBP}
+----
+
+To provide a non-singleton stream of values, use an _argument group_.
+Inside an E-expression, a group starts with the digraph `(:` similar to a macro
+invocation, but without a macro reference:
 
 ----
-(:prices [1, 2, 3] GBP) ⇒ {amount:1, currency:GBP}
+(:prices (:) GBP)       ⇒ _void_
+(:prices (: 1) GBP)     ⇒ {amount:1, currency:GBP}
+(:prices (: 1 2 3) GBP) ⇒ {amount:1, currency:GBP}
                           {amount:2, currency:GBP}
                           {amount:3, currency:GBP}
 ----
 
-Within the group, the invocation can have any number of arguments, including macro invocations.
+Within the group, the invocation can have any number of expressions that align
+with the parameter type.
 The macro parameter produces the results of those expressions, concatenated into a
 single stream, and the expander verifies that each value on that stream is acceptable by the
 parameter’s declared type.
 
 ----
-(:prices [1, (:values 2 3), 4] GBP) ⇒ {amount:1, currency:GBP}
+(:prices (: 1 (:values 2 3) 4) GBP) ⇒ {amount:1, currency:GBP}
                                       {amount:2, currency:GBP}
                                       {amount:3, currency:GBP}
                                       {amount:4, currency:GBP}
 ----
 
-[IMPORTANT]
-====
-To avoid ambiguity, the delimiter is required even for singleton values.  Consider this
-macro:
-
-[{nrm}]
-----
-(*macro* ouch [(stuff [*list*])] …)
-----
-
-Without this rule, the E-expression `(:ouch [])` would be ambiguous whether the parameter was
-intended to be void or a singleton empty-list value.
-====
-
-Grouping says whether multiple _arguments_ can be provided, while cardinality describes the
-number of _values_ those argument(s) must produce.
-The parameter declaration `(amount [*number*])` makes grouping explicit, with a default
-cardinality of zero-or-more.
-The declaration `(amount *[number]+*)` is also valid,
-indicating that the sequence of arguments must produce at least one value.
+Argument groups are allowed for parameters that do not have exactly-one cardinality.
 
-TIP: Grouped parameters cannot use the `*?*` and `*!*` modifiers; there's no point in
-requiring a grouping list when no more than one value is allowed.
+// TODO
+WARNING: The allowed combinations of cardinality and argument groups is pending
+finalization of the binary encoding.
 
-TIP: Rest parameters are effectively another grouping mode, so they cannot be combined with `[]`.
+TIP: Rest parameters effectively use implicit groups at the call site, and they cannot be
+invoked with explicit groups.
```

Comment:
Why not? We already have rules that define how to combine a stream in an expression group, allowing things like `(:prices (: 1 (:values 2 3) 4) GBP)`.

Is it legal to nest expression groups for non-"rest" parameters? E.g.: `(:prices (: 1 (: 2 3) 4) GBP)`

Reply:
I think your question is conflating expression groups and `(:values ...)`, which this PR hopes to make into distinct entities.

> Is it legal to nest expression groups for non-"rest" parameters? E.g.: `(:prices (: 1 (: 2 3) 4) GBP)`

No, this isn't legal; expression groups can only appear in argument positions. It's easier to understand why when you consider the binary encoding.  Say this is the definition of your macro (borrowed from the doc):

```
(macro prices
  [(amount number*),(currency symbol)]
  // ...BODY EXPR...
)
```

When the binary reader encounters the `amount` argument, it will know to decode an expression group and decode its contents: a sequence of tagged values that must be `number`s. There is no opcode for "expression group", so there's no way to encode `(: 2 3)` within the outer expression group.

I think we _could_ allow an explicit expression group for a rest parameter in text, but I don't think we want to.

>> Rest parameters effectively use implicit groups at the call site, and they cannot be
invoked with explicit groups.
> Why not? We already have rules that define how to combine a stream in an expression group, allowing things like (:prices (: 1 (:values 2 3) 4) GBP).

I think we _could_ permit it semantically, but it would be strictly redundant syntactically. As an example, let's flip `currency` and `amount` so the latter can be a rest parameter.

```
(macro prices
  [(currency symbol) (amount ...number)]
  // ...BODY EXPR...
)

(:prices USD 1 2 3 4)
(:prices USD (: 1 2 3 4)) // if we allowed "explicit rest"
```
We'd need to add an extra branch to support this, and I don't think it buys us much. In fact, I think that someone passing an explicit expression group to a `...` parameter is a helpful indication that they've misunderstood something and an opportunity to emit a helpful error message.

Reply:
@zslayton expresses my thinking well.  In particular, this is great:

> There is no opcode for "expression group"